File: Data-Structure-I/102-Binary-Tree-Level-Order-Traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> levelOrder(TreeNode* root) {
        vector<vector<int>> v;
        int depth = 0;
        dfs(root, depth, v);
        return v;
    }
    void dfs(TreeNode* node, int depth, vector<vector<int>>& v){
        if(node==nullptr) return;
        if(depth==v.size()) v.push_back({});
        v[depth].push_back(node->val);
        dfs(node->left, depth+1, v);
        dfs(node->right, depth+1, v);
    }
};

File: Data-Structure-I/102-Binary-Tree-Level-Order-Traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def levelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        tree, depth = [], 0
        self.dfs(root, 0, tree)
        return tree
    
    def dfs(self, node: Optional[TreeNode], depth: Optional[int], tree):
        if not node:
            return
        if depth >= len(tree):
            tree.append([])
        tree[depth].append(node.val)
        self.dfs(node.left, depth+1, tree)
        self.dfs(node.right, depth+1, tree)
        
        
